File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.java
import java.util.*;

class Solution {
    
    public int solution(String s) {
        
        int ans = Integer.MAX_VALUE;
        
        for(int len = 1; len <= s.length(); len++) {
            
            ans = Math.min(ans, compress(s, len));
        }
        
        return ans;
    }
    
    public int compress(String s, int len) {
        
        ArrayList<String> list = new ArrayList<>();
        
        for(int i = 0; i < s.length(); i += len) {
            int j = Math.min(i + len, s.length());
            
            list.add(s.substring(i, j));
        }
        
        StringBuilder sb = new StringBuilder();
        
        String prev = "";
        int count = 0;
        
        for(String str : list) {
            if(str.equals(prev)) {
                count++;
            } 
            else {
                
                if(count >= 2) {
                    sb.append(count);
                }
                
                sb.append(prev);
                prev = str;
                count = 1;
            }
        }
        
        if(count >= 2) {
            sb.append(count);
        }
        
        sb.append(prev);
        
        return sb.length();
    }
}

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.py
def compress(s, length):

    str_list = []

    for i in range(0, len(s), length):
        str_list.append(s[i:i + length])

    sb = []
    prev = ""
    count = 0

    for now in str_list:
        if now == prev:
            count += 1
        else:
            if count >= 2:
                sb.append(str(count))
            sb.append(prev)
            prev = now
            count = 1

    if count >= 2:
        sb.append(str(count))
    sb.append(prev)

    return len("".join(sb))

def solution(s):
    
    ans = float('inf')

    for length in range(1, len(s) + 1):
        ans = min(ans, compress(s, length))
    
    return ans